software/contrib/turing_machine.py

```diff
@@ -0,0 +1,254 @@
+        significant bit's position. This translates to the probability that the sequence will change. The
+        flip_probability is represented as an integer in the range [0, 100].
+        """
+        return self.flip_probability_getter()
+
+    @flip_probability.setter
+    def flip_probability(self, probability: int):
+        """Set the flip probability as an integer in the range [0, 100]."""
+        if probability < 0 or probability > 100:
+            raise ValueError(f"Probability of {probability} is outside the expected range of [0,100]")
+        self._flip_probability = probability
+
+    @property
+    def scale(self):
+        """Returns the current scaling factor, used to reduce the range of the output voltage to something lower than
+        the `max_output_voltage`. Represented by a float in the range [0, `max_output_voltage`]"""
+        return self.scale_getter()
+
+    @scale.setter
+    def scale(self, scale):
+        """Set the scale factor as a float in the range [0, `max_output_voltage`]"""
+        if scale < 0 or scale > self.max_output_voltage:
+            raise ValueError(f"Scale of {scale} is outside the expected range of [0,{self.max_output_voltage}]")
+        self._scale = scale
+
+    @property
+    def length(self):
+        """Returns the length of the current sequence as an integer in the range of [2, `bit_count`]"""
+        return self.length_getter()
+
+    @length.setter
+    def length(self, length):
+        """Sets the length of the current sequence as an integer in the range of [2, `bit_count`]"""
+        if length < 2 or length > self.bit_count:
+            raise ValueError(f"Length of {length} is outside the expected range of [2,{self.bit_count}]")
+        self._length = length
+
+    @property
+    def write(self):
+        """Returns the current value of the 'write switch'. When true the least significant bit will be cleared during
+        rotation, regardless of the `flip_probability`. This allows for real-time user manipulation of the sequence."""
+        return self.write_getter()
+
+    @write.setter
+    def write(self, value: bool):
+        """Set the state of the 'write switch'. `True` means that the least significant bit will be cleared."""
+        self._write = value
+
+
+class EuroPiTuringMachine(EuroPiScript):
+    def __init__(self, bit_count=DEFAULT_BIT_COUNT, max_output_voltage=MAX_OUTPUT_VOLTAGE):
+        self.tm = TuringMachine(bit_count, max_output_voltage)
+        self.tm.flip_probability_getter = self.flip_probability
+        self.tm.scale_getter = self.scale
+        self.tm.length_getter = self.length
+        self.tm.write_getter = self.write
+        self.tm.step_handler = self.step_handler
+        self.k2_scale_mode = True  # scale mode or length mode
+
+        @din.handler
+        def clock():
+            self.tm.step()
+
+        @b2.handler_falling
+        def mode_toggle():
+            if self.k2_scale_mode:
```

Comment:
You might want to consider a third mode where neither attribute is editable, which would help avoid accidentally moving the knob and changing a value unexpectedly.

Reply:
I didn't like this at first, but it will be necessary in order to load a saved state.

Reply:
added.